lib: Add PLL enable/disable to UsbBus impl

rmsyn opened the pull request:
Adds `PLL` enable/disable to `UsbBus` impl, fixing the TODO comments.

Bumps `avr-device` to `0.5.x` to have consistent re-exports with `avr-hal`.

Minor changes for API differences in `avr-device`.

rmsyn committed:
lib: add PLL enable/disable

Adds PLL enable and disable to UsbBus impl.

rmsyn committed:
examples: update example to pass the PLL register

Adds passing the PLL register into the USB struct for PLL enable/disable
in the UsbBus impl.

rmsyn committed:
bump avr-device to 0.5

Match the dependency used in `avr-hal` for consistent re-exports.

Fixes building the examples.

rmsyn committed:
fixup: lib: clear warnings about unnecessary unsafe

Remove `unsafe` sections in `uedatx` and `uenum` writers. Under the
`0.5.x` API, the `bits` function is no longer unsafe for those register
fields.

agausmann commented:
Thanks for the contribution!

I think this is a good time to explain why I haven't added PLL yet - I don't want this crate to require ownership of the PLL, because it can be shared with other peripherals (e.g. on the ATmega32u4, the high-speed timer) and the crate user may not want it to be turned off.

So, I probably will not merge this as-is, but I would accept a change to store an `Option<PLL>`. Or even better, a flexible API using a "notifier" trait so it can accept `()` (default constructor; do nothing), `PLL`, or a user-defined notifier type so they can conditionally enable/disable the PLL using their own logic.

agausmann commented on src/lib.rs:
```diff
@@ -381,14 +383,17 @@ impl usb_device::bus::UsbBus for UsbBus {
-            //TODO enable PLL
+            let pll = self.pll.borrow(cs);
+            pll.pllcsr
+                .modify(|_, w| w.pindiv().set_bit().plle().set_bit());
```
I think we need to wait for PLL lock here before accessing the USB registers.

Like from the example:

https://github.com/agausmann/atmega-usbd/blob/30fcf0398e9d225f086980898c34fc000c7a0a8c/examples/arduino_keyboard.rs#L47-L51

rmsyn replied:
Ok, I'll add that in with the other changes.

rmsyn commented:
> I think this is a good time to explain why I haven't added PLL yet - I don't want this crate to require ownership of the PLL, because it can be shared with other peripherals (e.g. on the ATmega32u4, the high-speed timer) and the crate user may not want it to be turned off.

I had concerns about this, too. I'll experiment with a less restrictive API.

> So, I probably will not merge this as-is, but I would accept a change to store an `Option<PLL>`. Or even better, a flexible API using a "notifier" trait so it can accept `()` (default constructor; do nothing), `PLL`, or a user-defined notifier type so they can conditionally enable/disable the PLL using their own logic.

Definitely, I'll try to get something working. I haven't done a "notifier" trait before. Do you have some examples of what you mean?

rmsyn committed:
fixup: review changes for optional PLL

Adds changes to not require PLL register for UsbBus.

When PLL is present, wait for PLL lock during USB enable before
accessing USB registers.

agausmann commented:
"Notifier trait" is not really a pattern that I'm aware of, I'm just giving the trait a name.

Basically, the crate would define a trait like this:

```rust
pub trait SuspendNotifier {
    fn suspend(&self);
    fn resume(&self);
}

impl SuspendNotifier for () {
    fn suspend(&self) {}
    fn resume(&self) {}
}

impl SuspendNotifier for PLL {
    fn suspend(&self) { /* disable PLL */ }
    fn resume(&self) { /* enable PLL, wait for lock */ }
}

```

Then `UsbBus` can have a generic type that implements `SuspendNotifier`, and take ownership of a value of that type. Then when `UsbBus::suspend` and `UsbBus::resume` is called, it can call the respective methods of the notifier.

This supports the following use cases:

- Safe, easy default (`UsbBus<()>`) - Doesn't modify the PLL, may have higher power consumption but universally works fine even if other peripherals require the PLL clock. (This should also be recommended by the documentation if power saving is not required)

- Easy power saving (`UsbBus<PLL>`) - Automatically enables/disables PLL, can be used when the USB peripheral is the only one using the PLL.

- Custom control (`UsbBus<CustomType>`) - Crate user can write custom logic to handle suspend/resume, so they can still safely power it down in the case where PLL is shared with other peripherals, they just need to provide that logic.

rmsyn committed:
lib: add SuspendNotifier trait

Adds a trait for a generic way to control suspend/resume notification.

The default implementation is a no-op, and the implementation around
`arduino_hal::pac::PLL` controls `PLLCSR` enabling/disabling.

agausmann commented:
Merged in as 5fafe22f442e696250e86fa536fd2b0b0f968aad
